### ISSUE TITLE
Qubes 4.0-rc2 Device Widget Ghost Entries

### ISSUE THREAD

**Issue by section0x8** [QubesOS/qubes-issues#3215]
The widget fails to keep a real time list of devices. If a device is not properly removed, the widget will not realize that it is gone. This can happen if the device is unplugged without detaching it. It can happen if a VM powers off without the device being removed first. And it can probably happen a number of other ways.

Screenshot:
![screenshot_2017-10-25_23-26-33](https://user-images.githubusercontent.com/18060570/32038046-f7685fac-b9dc-11e7-8b86-2c7e1f5aa8ea.png)

This goes without saying, I don't really understand how the widget organizes itself in the first place. There are two sets of rows what is the difference between them?

**Comment by marmarek** [comment]
> This goes without saying, I don't really understand how the widget organizes itself in the first place. There are two sets of rows what is the difference between them?

Those are different type/layers of devices - at the top are block devices, below are usb devices. If a device is both, it is in both places (but you can choose how you want to attach it).

**Comment by ERFier** [comment]
I could confirm this issue yesterday. It was also the case if one restarts the usb-vm for some reason. It's not the thumbdrive getting displayed as general usb device as well as block device but non blockdevices, like WWAN Card or Yubikey in my case, getting duplicated.

Today, I think due to some update I applied yesterday, I cant reproduce the behavior. But the widget still lacks a proper update. If I shutdown the usb-vm the devices are still shown despite the fact they are not accessible anymore. The devices are also still shown as connected and there is no way to detach them and reattach them as the usb-vm starts again or the device is plugged in again. Maybe a force update function accessible thru a right-click context menu is a good idea to at least empower the user to recover the widget once it displays nonsense.

**Comment by quaterniol** [comment]
When the menu for the USB device widget is open and an USB mouse (e.g. Logitech mouse) is removed, then almost always the entry for the mouse stays in the menu. Inserting the mouse again creates a new entry. Repeating this creates a lot of entries. (I am using qubes 4.0-rc2 and saw the entry disappear when the mouse was removed only once.)

**Comment by awokd** [comment]
In 4.0 rc3, clicking Eject on an attached USB device seems to do nothing. Selecting it again from the widget shows it still connected to the AppVM. Seems like it might be related to this issue, but should I open a new one?

**Comment by ThierryIT** [comment]
same for 4.0 rc4 and doesn't show anything in the Qubes manager has it was possible with the Qubes 3.2.

**Comment by andrewdavidwong** [comment]
> In 4.0 rc3, clicking Eject on an attached USB device seems to do nothing. Selecting it again from the widget shows it still connected to the AppVM. Seems like it might be related to this issue, but should I open a new one?

Sure, that sounds different enough to merit a separate issue. It could turn out to be the same cause, but we have no way of knowing that right now.

**Comment by RefinedSoftwareLLC** [comment]
Also, when the DVD in a drive is changed, the displayed title of the current DVD isn't updated.
